src/OpenLoco/Windows/IndustryList.cpp

```diff
@@ -914,39 +920,124 @@ namespace OpenLoco::Ui::Windows::IndustryList
+            {
+                return res;
+            }
+            _industryGhostPos = placementArgs.pos;
+            _industryGhostType = placementArgs.type;
+            _industryGhostId = _industryLastPlacedId;
+            _industryGhostPlaced = true;
+            return res;
+        }
+
+        // 0x0045857B
+        static std::optional<GameCommands::IndustryPlacementArgs> getIndustryPlacementArgsFromCursor(const int16_t x, const int16_t y)
+        {
+            auto* industryListWnd = WindowManager::find(WindowType::industryList);
+            if (industryListWnd == nullptr)
+            {
+                return {};
+            }
+
+            if (industryListWnd->current_tab != (Common::widx::tab_new_industry - Common::widx::tab_industry_list))
+            {
+                return {};
+            }
+
+            if (industryListWnd->row_hover == -1)
+            {
+                return {};
+            }
+
+            const auto pos = ViewportInteraction::getTileStartAtCursor({ x, y }); //ax,cx
+            if (!pos)
+            {
+                return {};
+            }
+
+            GameCommands::IndustryPlacementArgs args;
+            args.pos = *pos;
+            args.type = industryListWnd->row_hover; //dl
+            args.srand0 = _dword_E0C394;
+            args.srand1 = _dword_E0C398;
+            if (isEditorMode())
+            {
+                args.buildImmediately = true; //bh
+            }
+            return { args };
+        }
+
         // 0x0045848A
         static void onToolUpdate(Window& self, const WidgetIndex_t widgetIndex, int16_t x, const int16_t y)
         {
-            registers regs;
-            regs.esi = (int32_t)&self;
-            regs.dx = widgetIndex;
-            regs.ax = x;
-            regs.bx = y;
-            call(0x0045848A, regs);
+            Map::TileManager::mapInvalidateSelectionRect();
+            Input::resetMapSelectionFlag(Input::MapSelectionFlags::enable);
+            auto placementArgs = getIndustryPlacementArgsFromCursor(x, y);
+            if (!placementArgs)
+            {
+                removeIndustryGhost();
+                return;
+            }
+
+            Input::setMapSelectionFlags(Input::MapSelectionFlags::enable);
+            Map::TileManager::setMapSelectionCorner(4);
+            Map::TileManager::setMapSelectionArea(placementArgs->pos, placementArgs->pos);
+            Map::TileManager::mapInvalidateSelectionRect();
+
+            if (_industryGhostPlaced)
+            {
+                if (*_industryGhostPos == placementArgs->pos && _industryGhostType == placementArgs->type)
+                {
+                    return;
+                }
+            }
+
+            removeIndustryGhost();
+            auto cost = placeIndustryGhost(*placementArgs);
+            if (cost != dword_E0C39C)
+            {
+                dword_E0C39C = cost;
+                self.invalidate();
+            }
```

Comment:
I wonder if we can make the cost zero in sandbox mode as well. Not sure how it works in editor mode? I guess this is checked in the game command.